### PR TITLE
oracle: update NeoFS SDK to 1.0.0-rc.9

### LINKED ISSUE #3032: Update NeoFS SDK to RC9
It has some nice things for us to use, including `io.ReadCloser` returned from Get which can help in deduplicating some Oracle code.

### PR DESCRIPTION
**Pull request by AnnaShaleva**
Close #3032.

### CODE REVIEW

File: pkg/services/oracle/neofs/neofs.go [hunk 1]
```diff
@@ -47,7 +47,7 @@ var (
 // Get returns a neofs object from the provided url.
 // URI scheme is "neofs:<Container-ID>/<Object-ID/<Command>/<Params>".
 // If Command is not provided, full object is requested.
-func Get(ctx context.Context, priv *keys.PrivateKey, u *url.URL, addr string, resReader ResultReader) ([]byte, error) {
+func Get(ctx context.Context, priv *keys.PrivateKey, u *url.URL, addr string, resReader ResultReaderCloser) ([]byte, error) {
```
Comment (roman-khimov) [inline pkg/services/oracle/neofs/neofs.go:50]: I think you can return `io.ReadCloser()` here and drop `ResultReader` completely.

File: pkg/services/oracle/request.go [hunk 2]
```diff
@@ -169,7 +170,10 @@ func (o *Oracle) processRequest(priv *keys.PrivateKey, req request) error {
 			ctx, cancel := context.WithTimeout(context.Background(), o.MainCfg.NeoFS.Timeout)
 			defer cancel()
 			index := (int(req.ID) + incTx.attempts) % len(o.MainCfg.NeoFS.Nodes)
-			resp.Result, err = neofs.Get(ctx, priv, u, o.MainCfg.NeoFS.Nodes[index], readResponse)
+			resp.Result, err = neofs.Get(ctx, priv, u, o.MainCfg.NeoFS.Nodes[index], func(closer io.ReadCloser) ([]byte, error) {
+				defer closer.Close() // intentionally skip the closing error, make it unified with Oracle `https` protocol.
+				return readResponse(closer)
+			})
 			if err != nil {
 				if errors.Is(err, ErrResponseTooLarge) {
```
Comment (roman-khimov) [inline pkg/services/oracle/request.go:178]: This would then allow to deduplicate these handlers.

File: pkg/services/oracle/neofs/neofs.go [hunk 3]
```diff
@@ -60,30 +57,34 @@ func Get(ctx context.Context, priv *keys.PrivateKey, u *url.URL, addr string, re
...
 
 	var prmd client.PrmDial
 	prmd.SetServerURI(addr)
 	prmd.SetContext(ctx)
 	err = c.Dial(prmd) //nolint:contextcheck // contextcheck: Function `Dial->Balance->SendUnary->Init->setNeoFSAPIServer` should pass the context parameter
 	if err != nil {
-		return nil, err
+		return nil, nil, err
 	}
-	defer c.Close()
```
Comment (roman-khimov) [inline pkg/services/oracle/neofs/neofs.go:69]: OK, not really expected.

File: pkg/services/oracle/neofs/neofs.go [hunk 4]
```diff
@@ -47,10 +44,10 @@ var (
 // Get returns a neofs object from the provided url.
 // URI scheme is "neofs:<Container-ID>/<Object-ID/<Command>/<Params>".
 // If Command is not provided, full object is requested.
-func Get(ctx context.Context, priv *keys.PrivateKey, u *url.URL, addr string, resReader ResultReader) ([]byte, error) {
+func Get(ctx context.Context, priv *keys.PrivateKey, u *url.URL, addr string) (io.ReadCloser, func(), error) {
```
Comment (roman-khimov) [inline pkg/services/oracle/neofs/neofs.go:47]: But still not requiring an additional return value.

File: pkg/services/oracle/neofs/neofs.go [hunk 5]
```diff
@@ -60,30 +57,34 @@ func Get(ctx context.Context, priv *keys.PrivateKey, u *url.URL, addr string, re
...
-		return getHeader(ctx, c, objectAddr)
+		res, err = getHeader(ctx, c, objectAddr)
 	case ps[0] == hashCmd:
-		return getHash(ctx, c, objectAddr, ps[1:]...)
+		res, err = getHash(ctx, c, objectAddr, ps[1:]...)
 	default:
-		return nil, ErrInvalidCommand
+		err = ErrInvalidCommand
 	}
+	return res, finalizer, err
```
Comment (roman-khimov) [inline pkg/services/oracle/neofs/neofs.go:87]: With smth like
```
if err != nil || resBytes != nil {
    c.Close()
    if err != nil {
        return nil, err
    }
    return io.NopCloser(bytes.NewReader(resBytes)), nil
}
return clientCloseWrapper(res, c), nil
}

type clientCloseWrapper struct {
    io.ReadCloser
    c *client.Client
}
func (w clientCloseWrapper) Close() error {
    res := w.ReadCloser.Close()
    w.c.Close()
    return res
}
```

Comment (AnnaShaleva) [reply pkg/services/oracle/neofs/neofs.go]: I also thought about it, but to be honest, the previous code (the one we had before SDK update) looks better for me in this situation.

Fixed.

**Review by roman-khimov — APPROVED** [review]

**Pull request merged**